api/metric/api.go

```diff
@@ -16,66 +16,97 @@ package metric
 
 import (
 	"context"
+	"time"
 
 	"go.opentelemetry.io/api/core"
 	"go.opentelemetry.io/api/registry"
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
+//go:generate stringer -type=Kind
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid        Kind = iota
+	CumulativeKind      // Supports Add()
+	GaugeKind           // Supports Set()
+	MeasureKind         // Supports Record()
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Recorder interface {
+	// RecordAt allows the SDK to observe a single metric event
+	RecordAt(ctx context.Context, t time.Time, value float64)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
+type Meter interface {
+	// RecorderFor returns a handle for observing single measurements.
+	RecorderFor(context.Context, Instrument, ...core.KeyValue) Recorder
+
+	// RecordBatch atomically records a batch of measurements.
+	RecordBatch(context.Context, ...Measurement)
 }
 
-type Handle struct {
+// Instrument represents a named metric with recommended local-aggregation keys.
+type Instrument struct {
+	// Variable defines the metric name, description, and unit.
 	Variable registry.Variable
```

Comment:
Why is description and kind part of the key?

Reply:
This could maybe use some work. There is a common `Variable` type which represents metrics and keys, since they both are named things which we're interested in describing. 

There are some open questions about registration of key and metric names, to avoid passing around full structs. Is that what you're thinking?  I believe there is a desire to allow metric instruments to be allocated as vars in the compilation unit where they are used. As it stands, metrics and keys are not bound to the SDK, so that they do not need a SDK to initialize. They can initialize before any SDK exists, which is part of the reason why the Registry isn't well defined yet.

For my interest in a streaming SDK, I would be happy to record an event with a unique ID assigned to each key or metric, along with metadata like the description, and then at runtime the unique ID would stand in for the string that was registered. So, I'd like all keys and metric names to be assigned unique IDs, but don't want a registry to try and maintain a map of all keys/metrics in the process.